### ISSUE TITLE
Defragment framework interfaces into context based calls

### ISSUE THREAD

**Issue by fengjingchao** [go-distributed/meritop#107]
Problem Description
---
The latest build failure was caused in the race:

1. Everything worked correctly at epoch 6
2. After finished sending the data to parent and before new epoch set, a slave died. It restarted in a new node and repeated the same computation.
3. In regression framework,
```Go
func (t *dummySlave) ParentDataReady(parentID uint64, req string, resp []byte) {
        ...
	if len(children) != 0 {
		t.framework.FlagMetaToChild("ParamReady")
	} else {
		// On leaf node, we can immediately return by and flag parent
		// that this node is ready.
		t.framework.FlagMetaToParent("GradientReady")
	}
}
```Go
In FlagMetaToX, it prepended *epoch* to the string and set it in etcd. However, the current implementation is wrong about synchronizing epoch here:
```
func (f *framework) FlagMetaToChild(meta string) {
	value := fmt.Sprintf("%d-%s", f.epoch, meta)
	f.etcdClient.Set (ChildMetaPath, value, 0 )
}
```
The problem is, a new epoch 7 was set, and it flagged meta with the epoch 7. But it's actually dealing with data in epoch 6!

**Comment by fengjingchao** [comment]
@xiang90 @xiaoyunwu 
I am aware of the latest build failure and I have some thoughts on it. Help read and discuss it here. Let's define the context interface quickly as first step and then I can start implementing it.

**Comment by xiaoyunwu** [comment]
"In FlagMetaToX, it prepended epoch to the string and set it in etcd."
Why do we need to set that in etcd?

**Comment by fengjingchao** [comment]
@xiaoyunwu 
https://github.com/go-distributed/meritop/blob/7e62eb3c2879f84c231cd1ca5396c2f5232df06d/framework/bootstrap.go#L231

**Comment by fengjingchao** [comment]
More about the "epoch-meta":
1. At start-up, a new node will read meta from neighbors. 
2. But it didn't know whether the meta is flagged at current or previous epoch.
3. it parses epoch from "epoch-meta" and only passes the meta to user if epoch is current.

**Comment by xiaoyunwu** [comment]
ok, I got it, you mean the epoch is in meta, which is in etcd. 

But I still do not understand why we have issues in
func (f *framework) FlagMetaToChild(meta string) {
    value := fmt.Sprintf("%d-%s", f.epoch, meta)
    f.etcdClient.Set (ChildMetaPath, value, 0 )
}

Can you show a list of event that problem will arise? What if we prepend the epoch info to req in the framework, so that we know which epoch is this request about? Basically doing the same thing like we did to meta.

**Comment by fengjingchao** [comment]
Request has epoch. In spirit of #101 , we add epoch to all communication between frameworks. 

>  What if we prepend the epoch info to req in the framework, so that we know which epoch is this request about? 

This is user API. User didn't add epoch to the meta. We need to separate framework and task layers.

**Comment by xiaoyunwu** [comment]
I know, user call our framework with req, in our implementation we prepend epcoh, before we return result, we take out epcoh from req, that way, user do not have to know anything, but we have info we need to do the right thing.

**Comment by fengjingchao** [comment]
That's right. That's why I am proposing to give context to user and that context carries epoch, etc. info.

**Comment by xiaoyunwu** [comment]
but can we encode context into req then take it out?

**Comment by fengjingchao** [comment]
There are many choices to give context to user:
1. We can give context at SetEpoch if we assume this is only epoch specific context
2. We can give context at all callbacks?
3. We can also give context in other ways like encoding it into req.

It's decision making. I only care letting user call APIs in this way.

There is one more important thing to discuss: what should be called via Context not Framework? My list:
* FlagMetaToX
* DataRequest
* IncEpoch

**Comment by xiaoyunwu** [comment]
So u do not think pretend epoch would work?

**Comment by fengjingchao** [comment]
I didn't catch your question.. Could you use example?

**Comment by xiaoyunwu** [comment]
I think we should stay in simple if possible, what do u think?

**Comment by fengjingchao** [comment]
I agree simplicity is better. Do you have simpler solution for this?

**Comment by xiaoyunwu** [comment]
My preference is that we encoding the context into req and also response,
this way, user has one less thing to worry about. What do you think?

Otherwise, we have to again explain what is context, and how to use it.
Unless there are absolute reason for that, we should stay away.

On the separate note, I am wondering whether we can use the following
interface
ParentDataReady(parentID uint64, req string, resp ByteSink)
ChildDataReady(childID uint64, req string, resp ByteSink)

where ByteSink

type ByteSink interface{
    ReadAll(r io.Reader)
}

this way the application can get access to the raw data stream directly,
that can be a big saving when we copy gigabytes data around.

We can also make

// These are payload for application purpose.
ServeAsParent(fromID uint64, req string) []byte
ServeAsChild(fromID uint64, req string) []byte
use writer interface
ServeAsParent(fromId uint64, req string, io.Writer)
for the same reason.

Xiaoyun

**Comment by fengjingchao** [comment]
Let's talk with examples.

Let's assume we have encoded context into req:
```Go
func (t *dummySlave) ParentDataReady(parentID uint64, req Request, resp Resp) {
        ...
    if len(children) != 0 {
        req.Context.FlagMetaToChild("ParamReady")
    } else {
        req.Context.FlagMetaToParent("GradientReady")
    }
}
```

So my question is, what APIs should be included in Context not Framework? My list:

* FlagMetaToX
* DataRequest
* IncEpoch

**Comment by xiaoyunwu** [comment]
Is it possible to do away with the concept of context?

I still believe we can handle it without task implementation ever know
about context or epoch, am I missing something?

Xiaoyun

**Comment by fengjingchao** [comment]
See my **analysis** section and first paragraph in **solution** section. Without context, there is no way between framework and task to determine whether they are talking in the same epoch.

**Comment by xiaoyunwu** [comment]
Ok, Let me try to sequence it in a way that makes sense.

in epoch 6, task implementation says:
   t.framework.FlagMetaToChild("ParamReady")

before framework.FlagMetaToX executed, 
   epoch got move to epoch 7

in the function, the epoch 7 is encoded in FlagMetaToChild
func (f *framework) FlagMetaToChild(meta string) {
    value := fmt.Sprintf("%d-%s", f.epoch, meta)
    f.etcdClient.Set (ChildMetaPath, value, 0 )
}

I agree with you that instantiate a new context for each epoch is good idea. But can it be a parameter instead of a type.

**Comment by xiaoyunwu** [comment]
What is the reason that we should include incEpoch in context?

Here is my proposal:
on famework we add

GetContext() Context 

Where Context is a opaque type that includes epoch id. 

in Context it supports the follow three functions:
FlagMetaToParent(meta string)
FlagMetaToChild(meta string)
DataRequest 

What do you guys think?

**Comment by fengjingchao** [comment]
Adding GetContext() in framework is the same thing of FlagMetaToChild().

> What is the reason that we should include incEpoch in context?

I mentioned we need discussion on this. Currently I am fine only adding FlagMetaToX().

**Comment by xiaoyunwu** [comment]
What I really mean is following.

task implementation first get context from framework, that context then can
be used for the following three methods:
FlagMetaToX and DataRequest

Is this what you have in mind, if so, we are on the same page.

**Comment by fengjingchao** [comment]
No.

Task get **context** from each callback method. Because it's during this communication between them that causes epoch shift.

**Comment by xiaoyunwu** [comment]
Using a state pattern, let's introduce a state call Context. Slightly different from Hongchao's suggestion, I think we can get away with just state object with no method.

So I take it that we both agree that the following method in framework can be changed into 
	FlagMetaToParent(context Context, meta string)
	FlagMetaToChild(context Context, meta string)
	DataRequest(context Context, toID uint64, meta string)

As suggested by Hongchao, we can pass context as parameter to all the task implementation callback functions:
	// NOTE: the meta/data ready notifications follow at-least-once fault
	// tolerance semantics
	ParentMetaReady(context Context, parentID uint64, meta string)
	ChildMetaReady(context Context, childID uint64, meta string)
	ParentDataReady(context Context, parentID uint64, req string, resp []byte)
	ChildDataReady(context Context, childID uint64, req string, resp []byte)

The fundamental reason for the current test failure is that we need to have a state machine inside task implementation which handle its event in the right order. But I think add this context is a reasonable thing to do.
@hongchao,@xiang what do you think?

**Comment by fengjingchao** [comment]
It's too complicated to have APIs like this:
```Go
FlagMetaToParent(context Context, meta string)
FlagMetaToChild(context Context, meta string)
DataRequest(context Context, toID uint64, meta string)
```
Passing in a context will only adds burden of checking and complication.

**Comment by xiang90** [comment]
@fengjingchao How about using a generic context data struct? For example, an embed hash map? And we always attach Context to each request? 

Actually we can use this small lib for context and cancellation: 
http://blog.golang.org/context
http://godoc.org/golang.org/x/net/context

**Comment by xiang90** [comment]
/cc @xiaoyunwu 

Context is widely used in golang world. So we still can preserve the simplicity we want by not introducing "new" concepts. Actually this might also simplify the code by avoiding customized cancellation solution.
